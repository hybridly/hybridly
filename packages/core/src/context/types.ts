import type { Axios } from 'axios'
import type { Hooks } from '../plugins/hooks'
import type { Plugin } from '../plugins/plugin'
import type { RoutingConfiguration } from '../routing/types'
import type { PendingNavigation, ResolveComponent, SwapView, View, HybridPayload, Dialog } from '../router'

/** Options for creating a router context. */
export interface RouterContextOptions {
	/** The initial payload served by the browser. */
	payload: HybridPayload
	/** Adapter-specific functions. */
	adapter: Adapter
	/** History state serializer. */
	serializer?: Serializer
	/** List of plugins. */
	plugins?: Plugin[]
	/** The Axios instance. */
	axios?: Axios
	/** Initial routing configuration. */
	routing?: RoutingConfiguration
}

/** Router context. */
export interface InternalRouterContext {
	/** The current, normalized URL. */
	url: string
	/** The current view. */
	view: View
	/** The current, optional dialog. */
	dialog?: Dialog
	/** The current local asset version. */
	version: string
	/** The current adapter's functions. */
	adapter: ResolvedAdapter
	/** Scroll positions of the current page's DOM elements. */
	scrollRegions: ScrollRegion[]
	/** Arbitrary state. */
	state: Record<string, any>
	/** Currently pending navigation. */
	pendingNavigation?: PendingNavigation
	/** History state serializer. */
	serializer: Serializer
	/** List of plugins. */
	plugins: Plugin[]
	/** Global hooks. */
	hooks: Partial<Record<keyof Hooks, Array<Function>>>
	/** The Axios instance. */
	axios: Axios
	/** Routing configuration. */
	routing?: RoutingConfiguration
}

/** Router context. */
export type RouterContext = Readonly<InternalRouterContext>

/** Adapter-specific functions. */
export interface Adapter {
	/** Resolves a component from the given name. */
	resolveComponent: ResolveComponent
	/** Called when the view is swapped. */
	onViewSwap: SwapView
	/** Called when the context is updated. */
	onContextUpdate?: (context: InternalRouterContext) => void
	/** Called when a dialog is closed. */
	onDialogClose?: (context: InternalRouterContext) => void
}

export interface ResolvedAdapter extends Adapter{
	updateRoutingConfiguration: (routing?: RoutingConfiguration) => void
}

export interface ScrollRegion {
	top: number
	left: number
}

/** Provides methods to serialize the state into the history state. */
export interface Serializer {
	serialize: <T>(view: T) => any
	unserialize: <T>(state: any) => T
}

export interface SetContextOptions {
	/** Whether to propagate the context to adapters. */
	propagate?: boolean
}
