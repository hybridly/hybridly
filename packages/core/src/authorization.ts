export interface Authorizable<Authorizations extends Record<string, boolean>> {
	authorization: Authorizations
}

/**
 * Checks whether the given data has the authorization for the given action.
 * If the data object has no authorization definition corresponding to the given action, this method will return `false`.
 */
function can<
    Authorizations extends Record<string, boolean>,
    Data extends Authorizable<Authorizations>,
    Action extends keyof Data['authorization']
>(action: Action, resource: Data) {
    if (resource === undefined) {
        return useProperty('security.user').value?.authorization?.[action] ?? false
    }

    return resource.authorization?.[action] ?? false
}

function cannot<
    Authorizations extends Record<string, boolean>,
    Data extends Authorizable<Authorizations>,
    Action extends keyof Data['authorization']
>(action: Action, resource: Data) {
    return ! can(action, resource)
}
