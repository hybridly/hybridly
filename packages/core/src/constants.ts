export const STORAGE_EXTERNAL_KEY = 'hybridly:external'
export const HYBRIDLY_HEADER = 'x-hybrid'
export const EXTERNAL_NAVIGATION_HEADER = `${HYBRIDLY_HEADER}-external`
export const EXTERNAL_NAVIGATION_TARGET_HEADER = `${HYBRIDLY_HEADER}-external-target`
export const PARTIAL_COMPONENT_HEADER = `${HYBRIDLY_HEADER}-partial-component`
export const ONLY_DATA_HEADER = `${HYBRIDLY_HEADER}-only-data`
export const DIALOG_KEY_HEADER = `${HYBRIDLY_HEADER}-dialog-key`
export const DIALOG_REDIRECT_HEADER = `${HYBRIDLY_HEADER}-dialog-redirect`
export const EXCEPT_DATA_HEADER = `${HYBRIDLY_HEADER}-except-data`
export const VERSION_HEADER = `${HYBRIDLY_HEADER}-version`
export const ERROR_BAG_HEADER = `${HYBRIDLY_HEADER}-error-bag`
export const SCROLL_REGION_ATTRIBUTE = 'scroll-region'
