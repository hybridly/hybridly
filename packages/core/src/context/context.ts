import { debug } from '@hybridly/utils'
import axios from 'axios'
import { createSerializer } from '../router/history'
import { makeUrl } from '../url'
import type { HybridPayload } from '../router'
import { updateRoutingConfiguration } from '../routing'
import { runHooks } from '../plugins'
import type { RouterContext, InternalRouterContext, RouterContextOptions, SetContextOptions } from './types'

const state = {
	initialized: false,
	context: {} as InternalRouterContext,
}

/** Gets the current context. */
export function getRouterContext(): RouterContext {
	return getInternalRouterContext()
}

/** Gets the current context, but not in read-only. */
export function getInternalRouterContext(): InternalRouterContext {
	if (!state.initialized) {
		throw new Error('Hybridly is not initialized.')
	}

	return state.context
}

/** Initializes the context. */
export async function initializeContext(options: RouterContextOptions): Promise<InternalRouterContext> {
	state.initialized = true
	state.context = {
		...options.payload,
		responseErrorModals: options.responseErrorModals,
		serializer: createSerializer(options),
		url: makeUrl(options.payload.url).toString(),
		adapter: {
			...options.adapter,
			updateRoutingConfiguration,
		},
		scrollRegions: [],
		plugins: options.plugins ?? [],
		axios: options.axios ?? axios.create(),
		routing: options.routing,
		hooks: {},
		state: {},
	}

	await runHooks('initialized', {}, state.context)

	return getInternalRouterContext()
}

/**
 * Mutates properties at the top-level of the context.
 */
export function setContext(merge: Partial<InternalRouterContext> = {}, options: SetContextOptions = {}): void {
	Object.keys(merge).forEach((key) => {
		Reflect.set(state.context, key, merge[key as keyof InternalRouterContext])
	})

	if (options.propagate !== false) {
		state.context.adapter.onContextUpdate?.(state.context)
	}

	debug.context('Updated context:', { context: state.context, added: merge })
}

/** Gets a payload from the current context. */
export function payloadFromContext(): HybridPayload {
	return {
		url: getRouterContext().url,
		version: getRouterContext().version,
		view: getRouterContext().view,
		dialog: getRouterContext().dialog,
	}
}
