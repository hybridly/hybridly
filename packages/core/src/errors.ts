import type { AxiosResponse } from 'axios'

export class NotAHybridResponseError extends Error {
	constructor(public response: AxiosResponse) {
		super()
	}
}

export class NavigationCancelledError extends Error {}

export class RoutingNotInitialized extends Error {
	constructor() {
		super('Routing is not initialized. Make sure the Vite plugin is enabled and that `php artisan route:list` returns no error.')
	}
}

export class RouteNotFound extends Error {
	constructor(name: string) {
		super(`Route [${name}] does not exist.`)
	}
}

export class MissingRouteParameter extends Error {
	constructor(parameter: string, routeName: string) {
		super(`Parameter [${parameter}] is required for route [${routeName}].`)
	}
}
