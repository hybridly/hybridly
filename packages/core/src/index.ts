export { createRouter, router } from './router/router'
export type { Router, HybridRequestOptions, NavigationResponse, HybridPayload, ResolveComponent, Method, Progress } from './router/types'

export { getRouterContext } from './context'
export type { RouterContext, RouterContextOptions } from './context'

export { definePlugin, registerHook } from './plugins'
export type { Plugin } from './plugins'

export { makeUrl, sameUrls } from './url'
export type { UrlResolvable } from './url'

export { can } from './authorization'
export type { Authorizable } from './authorization'

export { route } from './routing'
export type { RouteDefinition, GlobalRouteCollection, RoutingConfiguration, RouteName, RouteParameters } from './routing'

export type { DynamicConfiguration } from './config'

export * as constants from './constants'
export * from './types'
