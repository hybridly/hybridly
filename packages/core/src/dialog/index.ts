import { getInternalRouterContext } from '../context'
import type { HybridRequestOptions } from '../router'
import { performLocalNavigation, performHybridNavigation } from '../router'

export interface CloseDialogOptions extends HybridRequestOptions {
	/**
	 * Close the dialog without a round-trip to the server.
	 * @default false
	 */
	local?: boolean
}

/**
 * Closes the dialog.
 */
export async function closeDialog(options?: CloseDialogOptions) {
	const context = getInternalRouterContext()
	const url = context.dialog?.redirectUrl ?? context.dialog?.baseUrl

	if (!url) {
		return
	}

	context.adapter.onDialogClose?.(context)

	if (options?.local === true) {
		return await performLocalNavigation(url, {
			preserveScroll: true,
			preserveState: true,
			dialog: false,
			component: context.view.component,
			properties: context.view.properties,
			...options,
		})
	}

	return await performHybridNavigation({
		url,
		preserveScroll: true,
		preserveState: true,
		...options,
	})
}
