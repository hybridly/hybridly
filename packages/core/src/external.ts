import type { AxiosResponse } from 'axios'
import qs from 'qs'
import { debug } from '@hybridly/utils'
import type { HybridRequestOptions } from 'hybridly'
import { EXTERNAL_NAVIGATION_HEADER, STORAGE_EXTERNAL_KEY } from './constants'
import { getRouterContext, setContext } from './context'
import { navigate } from './router/router'
import type { UrlResolvable } from './url'
import { makeUrl, sameUrls } from './url'

/**
 * Performs an external navigation by saving options to the storage and
 * making a full page reload. Upon loading, the navigation options
 * will be pulled and a hybrid navigation will be made.
 */
export async function performExternalNavigation(options: ExternalNavigationOptions): Promise<void> {
	debug.external('Navigating to an external URL:', options)
	window.sessionStorage.setItem(STORAGE_EXTERNAL_KEY, JSON.stringify(options))
	window.location.href = options.url

	// If the external navigation is to the same page, we need to manually perform
	// a full page reload
	if (sameUrls(window.location, options.url)) {
		debug.external('Manually reloading due to the external URL being the same.')
		window.location.reload()
	}
}

/** Navigates to the given URL without the hybrid protocol. */
export function navigateToExternalUrl(url: UrlResolvable, data?: HybridRequestOptions['data']) {
	document.location.href = makeUrl(url, {
		search: qs.stringify(data, {
			encodeValuesOnly: true,
			arrayFormat: 'brackets',
		}),
	}).toString()
}

/** Checks if the response wants to redirect to an external URL. */
export function isExternalResponse(response: AxiosResponse): boolean {
	return response?.status === 409 && !!response?.headers?.[EXTERNAL_NAVIGATION_HEADER]
}

/**
 * Performs the internal navigation when an external navigation to a hybrid view has been made.
 * This method is meant to be called on router creation.
 */
export async function handleExternalNavigation(): Promise<void> {
	debug.external('Handling an external navigation.')
	const options = JSON.parse(window.sessionStorage.getItem(STORAGE_EXTERNAL_KEY) || '{}') as ExternalNavigationOptions
	window.sessionStorage.removeItem(STORAGE_EXTERNAL_KEY)

	debug.external('Options from the session storage:', options)

	// If we navigated to somewhere with a hash, we need to update the context
	// to add said hash because it was initialized without it.
	setContext({
		url: makeUrl(getRouterContext().url, { hash: window.location.hash }).toString(),
	})

	await navigate({
		type: 'initial',
		preserveState: true,
		preserveScroll: options.preserveScroll,
	})
}

/** Checks if the navigation being initialized points to an external location. */
export function isExternalNavigation(): boolean {
	try {
		return window.sessionStorage.getItem(STORAGE_EXTERNAL_KEY) !== null
	} catch {}

	return false
}

interface ExternalNavigationOptions {
	/** Target URL. */
	url: string
	/** Whether to preserve the scroll if the external navigation leads to a hybrid view. */
	preserveScroll: boolean
}
