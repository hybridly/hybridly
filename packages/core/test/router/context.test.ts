import { expect, it } from 'vitest'
import { initializeContext, setContext } from '../../src/context'
import { fakeRouterContext, makeRouterContextOptions } from '../utils'

it('creates a valid router context', async() => {
	const context = await initializeContext(makeRouterContextOptions({
		payload: {
			url: 'https://localhost',
			version: '123',
			view: {
				component: 'my-owo-view',
				properties: {
					foo: 'bar',
				},
			},
		},
	}))

	expect(context).toMatchSnapshot('context')
})

it('updates the context', async() => {
	const context = await fakeRouterContext({
		payload: {
			url: 'https://localhost',
		},
	})

	expect(context).toMatchSnapshot('base context')

	setContext({
		url: 'https://localhost/foo/bar',
		version: 'new-version-string',
	})

	expect(context).toMatchSnapshot('updated context')

	setContext({
		url: 'https://localhost/foo/bar/dialog',
		version: 'new-version-string',
		dialog: {
			component: 'some.dialog',
			properties: {
				foo: 'bar',
			},
		},
	})

	expect(context).toMatchSnapshot('updated context with dialog')

	setContext({
		url: 'https://localhost/foo',
		version: 'new-version-string',
		dialog: undefined,
	})

	expect(context).toMatchSnapshot('updated context with no dialog')
})
