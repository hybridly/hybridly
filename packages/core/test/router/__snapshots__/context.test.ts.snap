// Vitest Snapshot v1

exports[`creates a valid router context > context 1`] = `
{
  "adapter": {
    "resolveComponent": [Function],
    "swapDialog": [Function],
    "swapView": [Function],
    "updateRoutingConfiguration": [Function],
  },
  "axios": [Function],
  "hooks": {},
  "plugins": [],
  "routing": undefined,
  "scrollRegions": [],
  "serializer": {
    "serialize": [Function],
    "unserialize": [Function],
  },
  "state": {},
  "url": "https://localhost/",
  "version": "123",
  "view": {
    "component": "my-owo-view",
    "properties": {
      "foo": "bar",
    },
  },
}
`;

exports[`updates the context > base context 1`] = `
{
  "adapter": {
    "resolveComponent": [Function],
    "swapDialog": [Function],
    "swapView": [Function],
    "updateRoutingConfiguration": [Function],
  },
  "axios": [Function],
  "hooks": {},
  "plugins": [],
  "routing": undefined,
  "scrollRegions": [],
  "serializer": {
    "serialize": [Function],
    "unserialize": [Function],
  },
  "state": {},
  "url": "https://localhost/",
  "version": "abc123",
  "view": {
    "component": "default.view",
    "properties": {},
  },
}
`;

exports[`updates the context > updated context 1`] = `
{
  "adapter": {
    "resolveComponent": [Function],
    "swapDialog": [Function],
    "swapView": [Function],
    "updateRoutingConfiguration": [Function],
  },
  "axios": [Function],
  "hooks": {},
  "plugins": [],
  "routing": undefined,
  "scrollRegions": [],
  "serializer": {
    "serialize": [Function],
    "unserialize": [Function],
  },
  "state": {},
  "url": "https://localhost/foo/bar",
  "version": "new-version-string",
  "view": {
    "component": "default.view",
    "properties": {},
  },
}
`;

exports[`updates the context > updated context with dialog 1`] = `
{
  "adapter": {
    "resolveComponent": [Function],
    "swapDialog": [Function],
    "swapView": [Function],
    "updateRoutingConfiguration": [Function],
  },
  "axios": [Function],
  "dialog": {
    "baseUrl": "https://base.test",
    "component": "some.dialog",
    "properties": {
      "foo": "bar",
    },
    "redirectUrl": "https://base.test/dialog",
  },
  "hooks": {},
  "plugins": [],
  "routing": undefined,
  "scrollRegions": [],
  "serializer": {
    "serialize": [Function],
    "unserialize": [Function],
  },
  "state": {},
  "url": "https://localhost/foo/bar/dialog",
  "version": "new-version-string",
  "view": {
    "component": "default.view",
    "properties": {},
  },
}
`;

exports[`updates the context > updated context with no dialog 1`] = `
{
  "adapter": {
    "resolveComponent": [Function],
    "swapDialog": [Function],
    "swapView": [Function],
    "updateRoutingConfiguration": [Function],
  },
  "axios": [Function],
  "dialog": undefined,
  "hooks": {},
  "plugins": [],
  "routing": undefined,
  "scrollRegions": [],
  "serializer": {
    "serialize": [Function],
    "unserialize": [Function],
  },
  "state": {},
  "url": "https://localhost/foo",
  "version": "new-version-string",
  "view": {
    "component": "default.view",
    "properties": {},
  },
}
`;
